modules/percolator/src/main/java/org/elasticsearch/percolator/ExtractQueryTermsService.java

```diff
@@ -153,33 +178,45 @@ public static void extractQueryTerms(Query query, ParseContext.Document document
                         continue;
                     }
 
-                    Set<Term> temp;
+                    Result temp;
                     try {
                         temp = extractQueryTerms(clause.getQuery());
                     } catch (UnsupportedQueryException e) {
                         uqe = e;
                         continue;
                     }
-                    bestClause = selectTermListWithTheLongestShortestTerm(temp, bestClause);
+                    bestClause = selectTermListWithTheLongestShortestTerm(temp.terms, bestClause);
                 }
                 if (bestClause != null) {
-                    return bestClause;
+                    // 1 required clause, no prohibited clause, maybe some optional clauses, so it is verified:
+                    boolean verified = mnsm <= 1 && numProhibitedClauses == 0 && numRequiredClauses == 1;
+                    return new Result(verified, bestClause);
                 } else {
                     if (uqe != null) {
                         throw uqe;
                     }
-                    return Collections.emptySet();
+                    return new Result(true);
                 }
             } else {
+                boolean verified = true;
                 Set<Term> terms = new HashSet<>();
                 for (BooleanClause clause : clauses) {
                     if (clause.isProhibited()) {
                         // we don't need to remember the things that do *not* match...
                         continue;
                     }
-                    terms.addAll(extractQueryTerms(clause.getQuery()));
+                    Result subResult = extractQueryTerms(clause.getQuery());
+                    if (subResult.canIgnore()) {
```

Comment:
I added this because so that MatchNoDocsQuery as part of bigger queries are just ignored and if they stand on their own we never mistake it for a match.